pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+func (cs *Coscheduling) sendStartScheduleSignal(pgName string) {
+	cs.pgReadyChan <- pgName
+	klog.Infof("Send StartScheduleSignal for %v success", pgName)
 }
 
-// markPodGroupAsExpired set the deletionTimestamp of PodGroup to mark PodGroup as expired.
-func (cs *Coscheduling) markPodGroupAsExpired(obj interface{}) {
-	pod := obj.(*v1.Pod)
-	podGroupName, podMinAvailable, _ := GetPodGroupLabels(pod)
-	if len(podGroupName) == 0 || podMinAvailable == 0 {
+// pgAdded reacts to a PG creation
+func (cs *Coscheduling) pgAdded(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Error(err)
 		return
 	}
-
-	pgKey := fmt.Sprintf("%v/%v", pod.Namespace, podGroupName)
-	// If it's a PodGroup and present in PodGroupInfos, set its deletionTimestamp.
-	value, exist := cs.podGroupInfos.Load(pgKey)
-	if !exist {
+	pg := obj.(*v1alpha1.PodGroup)
+	if pg.Status.Phase == v1alpha1.PodGroupFinished || pg.Status.Phase == v1alpha1.PodGroupFailed {
 		return
 	}
-	pgInfo := value.(*PodGroupInfo)
-	if pgInfo.deletionTimestamp == nil {
-		now := cs.clock.Now()
-		pgInfo.deletionTimestamp = &now
-		cs.podGroupInfos.Store(pgKey, pgInfo)
+	pgs := cs.pgMgr.PodGroupStatusCache().Get(key)
+	if pgs == nil {
+		klog.Infof("Add pg %v", key)
+		wait := util.GetWaitTimeDuration(pg, cs.scheduleTimeout)
+		pgs = cs.initPodGroupSchedulingStatus(pg, key, wait)
+		cs.pgMgr.PodGroupStatusCache().Set(key, pgs)
+		return
 	}
+	pgs.PodGroup = pg
 }
 
-// responsibleForPod selects pod that belongs to a PodGroup.
-func responsibleForPod(pod *v1.Pod) bool {
-	podGroupName, podMinAvailable, _ := GetPodGroupLabels(pod)
-	if len(podGroupName) == 0 || podMinAvailable == 0 {
-		return false
+// pgUpdated reacts to a PG update
+func (cs *Coscheduling) pgUpdated(_, new interface{}) {
+	cs.pgAdded(new)
+}
+
+// pgDelete reacts to a PG update
+func (cs *Coscheduling) pgDelete(new interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(new)
+	if err != nil {
+		klog.Error(err)
+		return
 	}
-	return true
+	cs.pgMgr.PodGroupStatusCache().Delete(key)
 }
 
-func (cs *Coscheduling) podGroupInfoGC() {
-	cs.podGroupInfos.Range(func(key, value interface{}) bool {
-		pgInfo := value.(*PodGroupInfo)
-		if pgInfo.deletionTimestamp != nil && pgInfo.deletionTimestamp.Add(time.Duration(*cs.args.PodGroupExpirationTimeSeconds)*time.Second).Before(cs.clock.Now()) {
-			klog.V(3).Infof("%v is out of date and has been deleted in PodGroup GC", key)
-			cs.podGroupInfos.Delete(key)
+// initPodGroupSchedulingStatus init pod groups
+func (cs *Coscheduling) initPodGroupSchedulingStatus(pg *v1alpha1.PodGroup, key string, expireTime time.Duration) *schecache.PodGroupSchedulingStatus {
+	pgs := &schecache.PodGroupSchedulingStatus{
+		PodGroup:         pg,
+		AssumedPodToNode: gochache.New(expireTime, 2*expireTime),
+		AssumedPods:      gochache.New(expireTime, 2*expireTime),
+	}
+	pgs.AssumedPods.OnEvicted(func(s string, i interface{}) {
```

Comment:
I suppose this is triggered when an entry (podUID) is deleted (L321) in StartBatchSchedule? By that time already allow() a waiting Pod (L320). So why we're rejecting all pods (L403) while also allowing the pod?

Reply:
as I  mentioned here https://github.com/kubernetes-sigs/scheduler-plugins/pull/57#discussion_r497216226, we should move the delete logic out of iterator. So here we would only iterator and `Allow`, when finishing iterator, we call `pendingPods.Delete(firstPodUid)`. We can not find pods in `waintingPods` of framework and would not reject any pod.